src/actions/mover.ts

```diff
@@ -218,15 +218,27 @@ export class Mover {
   /**
    * Get the source block for the cursor location, or undefined if no
    * source block can be found.
+   * If the cursor is on a shadow block, walks up the tree until it finds
+   * a non-shadow block to drag.
    *
    * @param workspace The workspace to inspect for a cursor.
    * @returns The source block, or undefined if no appropriate block
    *     could be found.
    */
   protected getCurrentBlock(workspace: WorkspaceSvg): BlockSvg | undefined {
-    const cursor = workspace?.getCursor();
-    const curNode = cursor?.getCurNode();
-    return (curNode?.getSourceBlock() as BlockSvg) ?? undefined;
+    const curNode = workspace?.getCursor()?.getCurNode();
+    let block = curNode?.getSourceBlock();
+    if (!block) return undefined;
+    while (block && block.isShadow()) {
```

Comment:
Nit: Because we check `!block` before and in the loop, this test can just be `block.isShadow()`. Makes it a bit clearer that block is always defined after the loop.

Reply:
It can return `null` from `getParent`. It'll be an error (see below), but still possible in the type system.